### PR TITLE
Bluetooth: Separate ISO from AUDIO and HCI HOST only

### PR DESCRIPTION
**Pull request by Thalley**
This PR has been updated to not only move the ISO Kconfig options, but the entire ISO implementation. The reasons for this is that BT_ISO Kconfigs should be available for both HCI Host and HCI Raw builds, and that ISO should also be available for other than BT_AUDIO builds.

### CODE REVIEW

**Comment by jhedberg** [comment]
Shouldn't there instead be some build-time condition for the controller-side ISO support?

**Comment by Thalley** [comment]
> Shouldn't there instead be some build-time condition for the controller-side ISO support?

As we briefly discussed in Slack, this file is really neither host nor controller. If anything, we should have some sort of common ISO condition. The scenario where I needed to remove this, was with a non-zephyr controller, but where I was using the rpmsg, so a Zephyr controller ISO config wouldn't really have an effect here either. 

I'm very open to other ideas/solutions.

**Comment by joerchan** [comment]
Kconfig symbols that are used in both are defined outside of both the controller and host folder, at the moment audio is inside of host, which is why we end up not having this definable for a controller-only build.
Maybe the audio kconfig symbols could be moved out of host?

**Comment by Thalley** [comment]
One of the other CONFIGs used in this file is `BT_RX_BUF_COUNT` which is also defined by the host, but is just not guarded by `BT_HCI_HOST` which ISO is. 

One solution is to move the line that sources the Audio Kconfig so that it doesn't require BT_HCI_HOST to be enabled (which is in fact the real problem here)

**Comment by jhedberg** [comment]
@Thalley hci_raw.c is never built with the host (as you noticed they're mutually exclusive). It's always built with an HCI driver however, which may be a native controller. I'm pretty sure we have shared Kconfig variables for other things, so seems like these should be moved there.

**Comment by jhedberg** [comment]
Just to follow up on my previous comment: enabling BT_HCI_HOST is not the right solution since hci_raw.c is a replacement for the host and should never be built together with it. Instead whatever Kconfig symbols are needed in hci_raw.c should be moved to some place that's independent of the host. Would it be less confusing to move hci_raw.c out from `subsys/bluetooth/host/` to some other place? It's been in the host directory since it's a replacement for the host (i.e. either you pick hci_raw.c from `host/` or everything else (the "real" host))

**Comment by jhedberg** [comment]
I also have a feeling that we have some confusion of terms here. When I talk about "the host" I'm talking about the full host stack in `subsys/bluetooth/host`. For historical reasons `hci_raw.c` is also in that folder but not part of the host stack. That's also why its Kconfig symbol is under `host/` and why some things there don't depend on `CONFIG_BT_HOST`.

**Comment by Thalley** [comment]
> I also have a feeling that we have some confusion of terms here. When I talk about "the host" I'm talking about the full host stack in `subsys/bluetooth/host`. For historical reasons `hci_raw.c` is also in that folder but not part of the host stack. That's also why its Kconfig symbol is under `host/` and why some things there don't depend on `CONFIG_BT_HOST`.

I think we might have been mixing the terms indeed - I've been referring to the host as either the BT_HCI_HOST or the BT_HCI_RAW :) 

That being said, I think it makes sense to move any such configs to a common place (we already have a common Kconfig file for that). 

CONFIG_BT_ISO is a bit... Difficult to deal with here, as that has been defined as a hidden option, which is enabled by enabling the `BT_AUDIO_UNICAST` which also enables a bunch of other stuff. 

I'd like to get some input from @Vudentz here, as he designed it (the ISO configs) in the first place.

**Comment by Vudentz** [comment]
> I think we might have been mixing the terms indeed - I've been referring to the host as either the BT_HCI_HOST or the BT_HCI_RAW :)
> 
> That being said, I think it makes sense to move any such configs to a common place (we already have a common Kconfig file for that).
> 
> CONFIG_BT_ISO is a bit... Difficult to deal with here, as that has been defined as a hidden option, which is enabled by enabling the `BT_AUDIO_UNICAST` which also enables a bunch of other stuff.
> 
> I'd like to get some input from @Vudentz here, as he designed it (the ISO configs) in the first place.

We could possible move CONFIG_BT_ISO as that is an HCI feature, BT_AUDIO_UNICAST is more of a interface with the host stack itself so I don't think we will be needing it on the likes of hci_raw.c or does the controller actually reuse the same definitions of buffer from the host? At least for ACL it doesn't seems to be the case as the likes of BT_ACL_RX_COUNT are defined in host/Kconfig, perhaps just the BT_MAX_ISO_CONN might be useful for the controller.

**Comment by Thalley** [comment]
> > I think we might have been mixing the terms indeed - I've been referring to the host as either the BT_HCI_HOST or the BT_HCI_RAW :)
> > That being said, I think it makes sense to move any such configs to a common place (we already have a common Kconfig file for that).
> > CONFIG_BT_ISO is a bit... Difficult to deal with here, as that has been defined as a hidden option, which is enabled by enabling the `BT_AUDIO_UNICAST` which also enables a bunch of other stuff.
> > I'd like to get some input from @Vudentz here, as he designed it (the ISO configs) in the first place.
> 
> We could possible move CONFIG_BT_ISO as that is an HCI feature, BT_AUDIO_UNICAST is more of a interface with the host stack itself so I don't think we will be needing it on the likes of hci_raw.c or does the controller actually reuse the same definitions of buffer from the host? At least for ACL it doesn't seems to be the case as the likes of BT_ACL_RX_COUNT are defined in host/Kconfig, perhaps just the BT_MAX_ISO_CONN might be useful for the controller.

@Vudentz The hci_raw.c/h files use Kconfigs from both the host and the controller for ACL, e.g. `BT_RX_BUF_COUNT` (host) and `BT_CTLR_TX_BUFFERS` (controller). 

I agree that we should not  enable `BT_AUDIO_UNICAST` for a HCI_RAW build, but we need to enable `BT_ISO` which is currently a hidden boolean. I guess we can just move that to a common Kconfig and "unhide", which uncouples it from Audio (which is arguable a good thing; I know at least one company intends to use ISO in Zephyr for something other than audio anyhow).

**Comment by joerchan** [comment]
Related to that is also: https://github.com/zephyrproject-rtos/zephyr/pull/30552

**Comment by Thalley** [comment]
@Vudentz @jhedberg @joerchan 

I've changed this PR to move the ISO configs to the common Kconfig for controller/host, and had the hci_raw.c guard the ISO buffers with `CONFIG_BT_ISO` again.

File: subsys/bluetooth/Kconfig
Comment (Vudentz) [inline subsys/bluetooth/Kconfig:233]: We might want to create something like BT_CIS then since MAX_CONN should only apply to CIS, that said the buffer definition might be useful for BIS as well.

Comment (Thalley) [reply subsys/bluetooth/Kconfig:233]: See https://github.com/zephyrproject-rtos/zephyr/pull/28409/files#r521374652
We can rename it to streams. What do you think?

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [inline subsys/bluetooth/Kconfig:243]: What "Isochronous TX buffers" is this referring to? HCI ISO Data packets? SDUs? PDUs? The terminology may have been clear on the host side, but it is in my opinion too fuzzy when being shared with the controller.

(also: Numer->Number)

Comment (Thalley) [reply subsys/bluetooth/Kconfig:243]: I fixed the numer->number part. For the other part, I agree that it is a bit fuzzy, and probably not something I can answer fully. My understanding of hci_raw.c is that it's partially host and partially controller. As far as I can tell, it's being used in the host for ISO Data packets (either over HCI or proprietary), but also for HCI in setups using raw_hci or rpmsg. 

@Vudentz  @jhedberg anything you'd like to add here?

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig]: Renamed to ISO_CHAN

**Comment by Thalley** [comment]
I updated the PR based on some build errors that was found if e.g. BT_AUDIO was not enabled.

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [inline subsys/bluetooth/Kconfig:251]: I assume this is referring to fragmenting SDUs in the host and not to fragmentation in ISOAL?

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [inline subsys/bluetooth/Kconfig:263]: Does this refer to the MTU of HCI Isochronous Data packets over HCI? How does it relate to ISO_Data_Packet_Length return parameter of the HCI_LE_Read_Buffer_Size command?

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [inline subsys/bluetooth/Kconfig:265]: Why is this limited to 2000 and not to, e.g., 4095, the maximum SDU length?

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [inline subsys/bluetooth/Kconfig:279]: Why is this limited to 2000 and not to, e.g., 4095, the maximum SDU length?

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [inline subsys/bluetooth/Kconfig:270]: I think this should refer to "HCI ISO Data RX buffers". The controller also has to deal with SDUs and PDUs, so the description should be explicit about this to avoid confusion what the setting refers to.

Also, Numer -> Number

**Review by wopu-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:235]: I agree. For the host, it doesn't matter if the ISO channel is for a BIS or a CIS.

**Comment by Thalley** [comment]
@wopu-ot You have a lot of good points w.r.t the ISO Kconfigs (though these were just moved in the PR). 

It does make me wonder if it makes more sense to move some of them back into the HCI_HOST Kconfigs instead? What we really need in the PR is actually just the HCI Buf count and the CONFIG_BT_ISO option. At this point (with ISO support in the controller in such an early state) I think it's difficult to define such Kconfigs that works for both the host and controller. What do you think?

**Comment by wopu-ot** [comment]
I think sharing the configuration as you do in this change is actually a good idea, as it minimizes possible mismatches between host and controller. I am just picky about the terminology because it needs to be more specific if there are more possible interpretations.

File: subsys/bluetooth/Kconfig
Comment (wopu-ot) [reply subsys/bluetooth/Kconfig:235]: Btw., in a discussion with @asmk-ot it came up that this is actually useful as it is for the ISOAL in the controller

**Comment by Thalley** [comment]
> I think sharing the configuration as you do in this change is actually a good idea, as it minimizes possible mismatches between host and controller. I am just picky about the terminology because it needs to be more specific if there are more possible interpretations.

Agreed. I'm not sure what the resulting terminology should be. E.g. the `BT_ISO_TX_BUF_COUNT` you comment on above is not just for HCI; it's used generically in the host and will be used like that in the host, even without using HCI for ISO data I suppose. I'd like to get some input from @Vudentz on these as he originally designed these.

File: subsys/bluetooth/Kconfig
Comment (Vudentz) [reply subsys/bluetooth/Kconfig]: I guess it quite obvious it is for ISO data packets as SDU, etc, are in the controller side but perhaps this will get more confusing if we put this as common Kconfig.

**Comment by Vudentz** [comment]
> > I think sharing the configuration as you do in this change is actually a good idea, as it minimizes possible mismatches between host and controller. I am just picky about the terminology because it needs to be more specific if there are more possible interpretations.
> 
> Agreed. I'm not sure what the resulting terminology should be. E.g. the `BT_ISO_TX_BUF_COUNT` you comment on above is not just for HCI; it's used generically in the host and will be used like that in the host, even without using HCI for ISO data I suppose. I'd like to get some input from @Vudentz on these as he originally designed these.

This was model after existing ACL Kconfig options, we can obviously change that if we want but until there is an actual ISO-AL it is hard to tell how useful these are going to be for the controller.

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:244]: As mentioned in another comment: This isn't just for HCI, but is also used in the host if ISO isn't done over HCI. 

As I understand it, it's used for full SDUs. If fragmentation is used/supported, the buffer count is controller by `CONFIG_BT_ISO_TX_FRAG_COUNT`. 

I updated the `help` message to clarify SDUs.

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:251]: Correct

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:263]: It is somewhat related to the length returned by the HCI_LE_Read_Buffer_Size command:

```
	max_num = MIN(rp->iso_max_pkt, CONFIG_BT_ISO_TX_BUF_COUNT);
	k_sem_init(&bt_dev.le.iso_pkts, max_num, max_num);
```

The `bt_dev.le.iso_pkts` is initialized the smallest of the two. How that affects the host is outside of my understanding of how this is implemented. Perhaps @Vudentz  can further explain this.

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:265]: It should be 4095. I'll update that.

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:279]: Updated

File: subsys/bluetooth/Kconfig
Comment (Thalley) [reply subsys/bluetooth/Kconfig:270]: Updated to specify SDUs

**Review by Vudentz — APPROVED** [review]
LGTM

**Review by wopu-ot — APPROVED** [review]

**Comment by Thalley** [comment]
Ping @asbjornsabo  as you probably need to be in the loop of this PR to do rebase after merging.

**Comment by Thalley** [comment]
@joerchan @jhedberg Can this be merged into the 2.5 release? Basically it's just fixing some ISO bugs that would be quite nice to have in the next release.

**Comment by jhedberg** [comment]
> @joerchan @jhedberg Can this be merged into the 2.5 release? Basically it's just fixing some ISO bugs that would be quite nice to have in the next release. 

That's a question for @nashif  . I have no objections.

**Review by jhedberg — APPROVED** [review]

**Comment by Thalley** [comment]
@nashif Can we merge this, or do you prefer waiting to after the release?

**Pull request merged**
